### PR TITLE
PWN-6396 - make RecipientDataBase and use it for sending

### PR DESCRIPTION
**Pull request by kabramovp2p**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-6396

## Description of Work

Make RecipientDataBase and use it for sending

## Screenshot

Android Emulator - Resizable_API_33:5554 2022-12-21 13-48-39

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/newsend/ui/search/NewSearchPresenter.kt [hunk 1]
```diff
@@ -30,18 +31,25 @@ class NewSearchPresenter(
...
-            // TODO PWN-6077 check latest recipients and show if exists
-            view?.showEmptyState(isEmpty = true)
-        } else {
-            val searchedItem = data.first()
-            val value = (searchedItem as? SearchResult.UsernameFound)?.username
-                ?: searchedItem.addressState.address
-            view?.showSearchValue(lastQuery ?: value)
-            setResult(data)
+        launch {
+            val data = lastResult.takeIf { lastResult.isNotEmpty() } ?: usernames
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/ui/search/NewSearchPresenter.kt:38]: what is data again? :))

File: app/src/main/java/org/p2p/wallet/send/SendModule.kt [hunk 2]
```diff
@@ -61,4 +64,8 @@ object SendModule : InjectionModule {
         factoryOf(::NewSendPresenter) bind NewSendContract.Presenter::class
         factoryOf(::NewSendDetailsPresenter) bind NewSendDetailsContract.Presenter::class
     }
+
+    private fun Module.initDataLayer() {
+        factoryOf(::RecipientsDatabaseRepository)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/send/SendModule.kt:69]: interface needed

File: app/src/main/java/org/p2p/wallet/send/model/SearchResult.kt [hunk 3]
```diff
@@ -13,11 +14,16 @@ sealed class SearchResult(open val addressState: AddressState) : Parcelable {
...
+        val sourceToken: Token.Active? = null,
+        val date: Date? = null
     ) : SearchResult(addressState)
 
     @Parcelize
-    data class UsernameFound(override val addressState: AddressState, val username: String) : SearchResult(addressState)
+    data class UsernameFound(
+        override val addressState: AddressState,
+        val username: String,
+        val date: Date? = null
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/send/model/SearchResult.kt:25]: dateFound?

File: app/src/main/java/org/p2p/wallet/utils/DateTimeUtils.kt [hunk 4]
```diff
@@ -68,6 +69,20 @@ object DateTimeUtils {
         }
     }
 
+    fun getDateRelatedFormatted(datetime: Long, context: Context): String {
+        val currentTime = System.currentTimeMillis()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/utils/DateTimeUtils.kt:73]: can be done with Duration too, but ok

File: app/src/main/java/org/p2p/wallet/newsend/db/RecipientDao.kt [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    suspend fun insertOrReplace(entity: RecipientEntry)
+
+    @Query(
+        """
+            UPDATE recipient_table
+            SET nickname = :nickname, date_timestamp = :dateTimestamp
+            WHERE address = :address 
+        """
+    )
+    suspend fun update(
```
[the remark this answers is not in the document]
Comment (kabramovp2p) [reply app/src/main/java/org/p2p/wallet/newsend/db/RecipientDao.kt:41]: it is done with insertOrReplace

[the remark this answers is not in the document]
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/newsend/db/RecipientDao.kt:41]: so why not use insertOrReplace?

File: app/src/main/java/org/p2p/wallet/newsend/ui/search/NewSearchPresenter.kt (hunk 1 above)
Comment (kabramovp2p) [reply app/src/main/java/org/p2p/wallet/newsend/ui/search/NewSearchPresenter.kt:38]: data is data =)

File: app/src/main/java/org/p2p/wallet/send/SendModule.kt (hunk 2 above)
Comment (kabramovp2p) [reply app/src/main/java/org/p2p/wallet/send/SendModule.kt:69]: for what?

File: app/src/main/java/org/p2p/wallet/newsend/ui/search/NewSearchPresenter.kt (hunk 1 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/newsend/ui/search/NewSearchPresenter.kt:38]: finalSearchResult?

File: app/src/main/java/org/p2p/wallet/send/model/SearchResult.kt (hunk 3 above)
Comment (kabramovp2p) [reply app/src/main/java/org/p2p/wallet/send/model/SearchResult.kt:25]: it is date of Transaction so i prefer not to do something like dateFound

File: app/src/main/java/org/p2p/wallet/utils/DateTimeUtils.kt (hunk 4 above)
Comment (kabramovp2p) [reply app/src/main/java/org/p2p/wallet/utils/DateTimeUtils.kt:73]: Duration ? what how? show me =)

File: app/src/main/java/org/p2p/wallet/send/SendModule.kt (hunk 2 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/send/SendModule.kt:69]: for Repository
all our repositories have interfactes
e.c. org.p2p.wallet.history.repository.local.TransactionDetailsLocalRepository

File: app/src/main/java/org/p2p/wallet/send/model/SearchResult.kt (hunk 3 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/send/model/SearchResult.kt:25]: i can't really couple transaction date and search result, but okay..
so date -> transactionDate?

File: app/src/main/java/org/p2p/wallet/utils/DateTimeUtils.kt (hunk 4 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/utils/DateTimeUtils.kt:73]: System.currentTimeMillis().toDuration(DurationUnit.MILLISECONDS)
maybe you can find something useful in those methods

File: app/src/main/java/org/p2p/wallet/newsend/repository/RecipientsDatabaseRepository.kt [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    suspend fun getRecipients(): List<SearchResult> =
+        recipientDao.getRecipients().map { it.toSearchResult() }
+
+    suspend fun clear() {
+        recipientDao.clearAll()
+    }
+
+    private fun RecipientEntry.toSearchResult(): SearchResult = if (nickname.isNullOrEmpty()) SearchResult.AddressOnly(
```
[the remark this answers is not in the document]
Comment (kabramovp2p) [reply app/src/main/java/org/p2p/wallet/newsend/repository/RecipientsDatabaseRepository.kt:26]: @1sakov do we ned mapper?
i prefer make classes when we need to move bunch of code but here we have only 2 local methods which used only here

File: app/src/main/java/org/p2p/wallet/newsend/db/RecipientDao.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (kabramovp2p) [reply app/src/main/java/org/p2p/wallet/newsend/db/RecipientDao.kt:41]: hm ok will need to check how it works

[the remark this answers is not in the document]
Comment (kabramovp2p) [reply app/src/main/java/org/p2p/wallet/newsend/db/RecipientDao.kt:41]: heh it is already used i will just remove this update =))

File: app/src/main/java/org/p2p/wallet/send/model/SearchResult.kt (hunk 3 above)
Comment (kabramovp2p) [reply app/src/main/java/org/p2p/wallet/send/model/SearchResult.kt:25]: no
in context of search result there is no transactionDate

File: app/src/main/java/org/p2p/wallet/newsend/model/RecipientConverter.kt [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+import org.p2p.core.utils.toBigDecimalOrZero
+import org.p2p.wallet.home.db.TokenEntity
+import org.p2p.wallet.newsend.db.RecipientEntity
+import org.p2p.wallet.send.model.AddressState
+import org.p2p.wallet.send.model.SearchResult
+import java.util.Date
+
+private const val EMPTY_TIMESTAMP = 0L
+
+object RecipientConverter {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/model/RecipientConverter.kt:15]: class
naming is RecipientRepositoryMapper

File: app/src/main/java/org/p2p/wallet/newsend/model/RecipientConverter.kt [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+    fun fromDatabase(
+        entity: RecipientEntity
+    ): SearchResult = if (entity.nickname.isNullOrEmpty()) SearchResult.AddressOnly(
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/model/RecipientConverter.kt:19]: {}

File: app/src/main/java/org/p2p/wallet/newsend/model/RecipientConverter.kt [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        addressState = AddressState(entity.address),
+        date = Date(entity.dateTimestamp)
+    ) else SearchResult.UsernameFound(
+        addressState = AddressState(entity.address),
+        username = entity.nickname,
+        date = Date(entity.dateTimestamp)
+    )
+
+    fun toDatabase(searchResult: SearchResult, newDate: Date): RecipientEntity {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/model/RecipientConverter.kt:28]: toEntity
fromEntity

File: app/src/main/java/org/p2p/wallet/newsend/model/RecipientConverter.kt [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+object RecipientConverter {
+
+    fun fromDatabase(
+        entity: RecipientEntity
+    ): SearchResult = if (entity.nickname.isNullOrEmpty()) SearchResult.AddressOnly(
+        addressState = AddressState(entity.address),
+        date = Date(entity.dateTimestamp)
+    ) else SearchResult.UsernameFound(
+        addressState = AddressState(entity.address),
+        username = entity.nickname,
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/model/RecipientConverter.kt:24]: nickname -> username

File: app/src/main/java/org/p2p/wallet/newsend/model/RecipientConverter.kt (hunk 7 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/newsend/model/RecipientConverter.kt:15]: Screen Shot 2022-12-21 at 21 01 36

Comment (kabramovp2p) [reply app/src/main/java/org/p2p/wallet/newsend/model/RecipientConverter.kt:15]: image
no i see TokenConverter

Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/newsend/model/RecipientConverter.kt:15]: old naming, but okay

File: app/src/main/java/org/p2p/wallet/newsend/model/RecipientConverter.kt [hunk 11]
```diff
@@ -16,14 +11,16 @@ object RecipientConverter {
...
-    ): SearchResult = if (entity.nickname.isNullOrEmpty()) SearchResult.AddressOnly(
-        addressState = AddressState(entity.address),
-        date = Date(entity.dateTimestamp)
-    ) else SearchResult.UsernameFound(
-        addressState = AddressState(entity.address),
-        username = entity.nickname,
-        date = Date(entity.dateTimestamp)
-    )
+    ): SearchResult {
+        return if (entity.username.isNullOrEmpty()) SearchResult.AddressOnly(
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/model/RecipientConverter.kt:15]: brackets for if {}

**Review by gslevinkov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/newsend/ui/search/NewSearchPresenter.kt [hunk 12]
```diff
@@ -30,18 +31,25 @@ class NewSearchPresenter(
...
+                } else {
+                    setResult(recentRecipients, R.string.search_recently)
+                }
+            } else {
+                val searchedItem = finalSearchResult.first()
+                val value = (searchedItem as? SearchResult.UsernameFound)?.username
+                    ?: searchedItem.addressState.address
+                view?.showSearchValue(lastQuery ?: value)
+                setResult(finalSearchResult)
+            }
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/newsend/ui/search/NewSearchPresenter.kt:52]: this function is getting complex too. 
add todo to make it more optimized please

**Review by 1sakov — APPROVED** [review]
lgtm, minor recommendations

File: app/src/main/java/org/p2p/wallet/newsend/ui/search/NewSearchPresenter.kt (hunk 12 above)
Comment (kabramovp2p) [reply app/src/main/java/org/p2p/wallet/newsend/ui/search/NewSearchPresenter.kt:52]: Done

**Pull request merged**